### PR TITLE
refactor!: implicit tag in ZLayer concat (++) method

### LINKED ISSUE #9262: [ZIO 1.x] Unused implicit tag in ZLayer concat (++) method
```scala
final def ++[E1 >: E, RIn2, ROut1 >: ROut, ROut2](
  that: ZLayer[RIn2, E1, ROut2]
)(implicit ev: Has.Union[ROut1, ROut2], tag: Tag[ROut2]): ZLayer[RIn with RIn2, E1, ROut1 with ROut2] =
  self.zipWithPar(that)(ev.union)

implicit def HasHasUnion[R <: Has[_], R1 <: Has[_]: Tag]: Union[R, R1] =
  new Union[R, R1] {
    def union(r: R, r1: R1): R with R1 =
      r.union[R1](r1)
  }
```

That implicit `Tag` in `++` is not used and it can generate a lot of garbage esp. when using `zio-magic`.

### PR DESCRIPTION
**Pull request by asr2003**
Closes #9262 
/claim #9262

### CODE REVIEW

**Comment by asr2003** [comment]
It seems mima check have been failed due comparision of  1.x wit 2.x?

File: core/shared/src/main/scala/zio/ZLayer.scala [hunk 1]
```diff
@@ -58,7 +58,7 @@ sealed abstract class ZLayer[-RIn, +E, +ROut] { self =>
    */
   final def ++[E1 >: E, RIn2, ROut1 >: ROut, ROut2](
     that: ZLayer[RIn2, E1, ROut2]
-  )(implicit ev: Has.Union[ROut1, ROut2], tag: Tag[ROut2]): ZLayer[RIn with RIn2, E1, ROut1 with ROut2] =
+  )(implicit ev: Has.Union[ROut1, ROut2]): ZLayer[RIn with RIn2, E1, ROut1 with ROut2] =
```
Comment (joroKr21) [inline core/shared/src/main/scala/zio/ZLayer.scala]: You need to add another method with the old JVM signature to appease MiMa

Comment (asr2003) [reply core/shared/src/main/scala/zio/ZLayer.scala]: Thanks!

Instead I introduced new method concat to make mima happier and for backward compatibility

**Comment by kyri-petrou** [comment]
I might be missing something here, but I don't see how this is resolving the issue. Adding a new method that does almost the same thing is quite bad for UX and requires a lot of code changes to get the benefit out of

**Comment by asr2003** [comment]
@kyri-petrou The mima is failing if i considered with one method
https://github.com/zio/zio/actions/runs/11745998347/job/32724715976#step:5:271

If breaking needed. I can revert to modify original one. And to overload methods the Tag is implicit param and causes ambiguity

**Comment by kyri-petrou** [comment]
> @kyri-petrou The mima is failing if i considered with one method

@asr2003 then I think the PR should figure out a way to modify the current method while maintaining binary compatibility. The `++` method on a `ZLayer` is _very_ widely used and we cannot afford deprecating it or breaking binary compatibility.

It might not be possible although I have a feeling there are some techniques that can be employed here to achieve this.

**Comment by asr2003** [comment]
@kyri-petrou I am thinking something like this, we can make `Tag[ROut2]` an optional implicit parameter by providing a default `null` value. This approach may avoid breaking binary compatibility and should satisfy MiMa checks and also existing functionality remains intact without requiring an additional method like concat.

**Comment by kyri-petrou** [comment]
> @kyri-petrou I am thinking something like this, we can make `Tag[ROut2]` an optional implicit parameter by providing a default `null` value.

I have a feeling that this won't work, at least not without changes to the Tag generation macros. If you're to try this approach please add a test or similar that showcases that the compiler doesn't generate the Tag when a default value is present (for all Scala versions)

**Comment by jdegoes** [comment]
One trick is to make the old method private, and then make the new method public.

**Comment by asr2003** [comment]
@jdegoes @kyri-petrou I have tried all the approaches we discussed and  it causing ambiguity of methods unless we change the method name. But in 2.x , `++`  method is already without `Tag` .

If deprecation is ok, lets go with it as in 2.x we have with no Tag

**Comment by kyri-petrou** [comment]
Closing in favour of https://github.com/zio/zio/pull/9297

**Pull request closed without merging**